### PR TITLE
Integrate EIA-860 2008 data

### PR DESCRIPTION
**Pull request by aesharpe**
Changes include: 

1. Added raw 2008 data into column maps, file maps, and header/footer maps.

2. Added a new raw table for pre-2009 generators. Instead of breaking it into existing, planned, and retired by by tab, they lump it into one csv with an `operational_status_code` column. Based on their designations (below) I made the assumption that **OP, SB, OA, OS = existing** and **RE = retired**. Proposed plants had their own file.

- OP | Operating - in service   (commercial operation) and producing some electricity.  Includes peaking units that are run on an   as-needed (intermittent or seasonal) basis.
- SB | Standby/Backup -   available for service but not normally used (has little or no generation   during the year).
- OA | Out of Service - was   not used for some or all of the reporting period but was either returned to   service on December 31 or will be returned to service in the next calendar   year.
- OS | Out of Service - was   not used for some or all of the reporting period and is not expected to be   returned to service in the next calendar year.
- RE | Retired - no longer in service and not expected to be returned   to service.

3. Updated the transform function for the generators table to accommodate this new aggregated raw table.

4. Removed an extraneous column from the column maps in the `plant` table (`ferc_exempt_wholesale_docket_no`).

**NOTE**: In the `plant` table, I noticed that the columns `street_address` and `po_box` don't always work out that way. That's to say, they behave more like `address_1` and `address_2` with the P.O. box sometimes appearing in the first column and other times in the second. I would love some feedback on whether to change the names of these columns or leave them as is. My vote would be to change them.

### CODE REVIEW

File: src/pudl/transform/eia860.py [hunk 1]
```diff
@@ -96,12 +99,21 @@ def generators(eia860_dfs, eia860_transformed_dfs):
...
     ge_df['operational_status'] = 'existing'
     gr_df['operational_status'] = 'retired'
+    g_df['operational_status'] = (
+        g_df['operational_status_code']
+        .replace({'OP': 'existing',  # could move this dict to constants
+                  'SB': 'existing',
+                  'OA': 'existing',
+                  'OS': 'existing',
+                  'RE': 'retired'})
+    )
```
Comment (zaneselvans) [inline src/pudl/transform/eia860.py:113]: I think it's fine here.

**Review by zaneselvans — COMMENTED** [review]
Hey overall this looks good. I agree that going to `address1` and `address2` rather than `street_address` and `po_box` makes sense if they aren't consistently showing up in the right places. What do you think @cmgosnell?

Note that the CI tests will only test the most recent year of data, so they won't ever catch errors in here. Have you run an all-years EIA ETL locally to see if it works from end to end?

**Comment by aesharpe** [comment]
I haven't run a test or tried out the harvesting process yet. Is there a pytest I can run or should I just recreate the functions?

**Comment by aesharpe** [comment]
> Hey overall this looks good. I agree that going to `address1` and `address2` rather than `street_address` and `po_box` makes sense if they aren't consistently showing up in the right places. What do you think @cmgosnell?
> 
> Note that the CI tests will only test the most recent year of data, so they won't ever catch errors in here. Have you run an all-years EIA ETL locally to see if it works from end to end?

I haven't run a test or tried out the harvesting process yet. Is there a pytest I can run or should I just recreate the functions?

**Comment by zaneselvans** [comment]
When I'm working integrating new years into an existing data source, I will typically create a new settings file for the `pudl_etl` script, which just tries to run the ETL for the data sources I'm working with (in this case `eia860` + `eia923` since they're tightly integrated) and then run `pudl_etl` against that settings file (making sure to give the output datapacakge a new name so it doesn't clobber my existing datapackage outputs).

Alternatively, you can run `pytest` without the `--fast` option, and edit the "full" settings file that's in the test settings to include the new data that you're integrating (the new year of eia860 data in this case), and then run the minimal `pytest` test required to exercise your new addition, which would be something like
```
pytest -sv --disable-warnings --pudl_in=AUTO --live_ferc1_db=AUTO test/etl_test.py::test_pudl_engine
```
This tells it just to run the `test_pudl_engine` test that's specified within the `etl_test.py` module, while using the already downloaded data in the local datastore, and while assuming that the FERC 1 DB has already been cloned.  However, since you don't really need to run the FERC 1 DB to test what you're working on right now, you might want to comment that out of the settings file... and that's part of why I usually just end up using the `pudl_etl` script with a temporary custom settings file.

**Comment by zaneselvans** [comment]
Ah, also, since there's a bunch of connections / dependencies between the `eia860` and `eia923` data, it's possible that you might want to get one year of each of them done, and then test them together. Though I think it should be the case that `eia923` depends on `eia860` and not vice-versa, so you could probably do all the `eia860` first and then move on to `eia923`

**Comment by zaneselvans** [comment]
Another thing we could maybe do is just concatenate the two address fields together into a single addressfield. Especially if one of them is often empty.

**Comment by aesharpe** [comment]
> Another thing we could maybe do is just concatenate the two address fields together into a single addressfield. Especially if one of them is often empty.

I like this idea

**Comment by aesharpe** [comment]
> Ah, also, since there's a bunch of connections / dependencies between the `eia860` and `eia923` data, it's possible that you might want to get one year of each of them done, and then test them together. Though I think it should be the case that `eia923` depends on `eia860` and not vice-versa, so you could probably do all the `eia860` first and then move on to `eia923`

I was thinking maybe it would work if I included old 923 and 860 data...? I'm not super clean on the dependencies, but I was under the impression that they might not be year-dependent. I'll run the etl now regardless.

File: src/pudl/constants.py [hunk 2]
```diff
@@ -1352,13 +1352,14 @@
...
+        ['utility_name_eia'],
         # annual cols
-        ['street_address', 'city', 'state', 'zip_code',
+        ['street_address', 'city', 'state', 'zip_code', 'entity_type',
          'plants_reported_owner', 'plants_reported_operator',
          'plants_reported_asset_manager', 'plants_reported_other_relationship',
-         ],
+         'utility_attn', 'address_2', 'zip_code_4', 'contact_firstname',
+         'contact_lastname', 'contact_title', 'contact_firstname2',
+         'contact_lastname2', 'contact_title2'],
```
[the remark this answers is not in the document]
Comment (aesharpe) [reply src/pudl/constants.py:1362]: good call

**Review by cmgosnell — APPROVED** [review]

File: src/pudl/transform/eia860.py [hunk 3]
```diff
@@ -96,12 +99,21 @@ def generators(eia860_dfs, eia860_transformed_dfs):
     gp_df = eia860_dfs['generator_proposed'].copy()
     ge_df = eia860_dfs['generator_existing'].copy()
     gr_df = eia860_dfs['generator_retired'].copy()
+    g_df = eia860_dfs['generator'].copy()
```
[the remark this answers is not in the document]
Comment (aesharpe) [reply src/pudl/transform/eia860.py:102]: I tried it without the 2008 and it seems to work fine. I think it's just a matter of where the function looks for the data to add to `generators_eia860` table.

File: src/pudl/package_data/meta/datapkg/datapackage.json [hunk 4]
```diff
@@ -5148,6 +5142,57 @@
...
                         "format": "default"
+                    },
+                    {
+                        "name": "entity_type",
+                        "type": "string",
+                        "description": "Entity type of principle owner (C = Cooperative, I = Investor-Owned Utility, Q = Independent Power Producer, M = Municipally-Owned Utility, P = Political Subdivision, F = Federally-Owned Utility, S = State-Owned Utility, IND = Industrial, COM = Commercial",
+                        "format": "default"
+                    },
+                    {
+                        "name": "utility_attn",
```
[the remark this answers is not in the document]
Comment (aesharpe) [reply src/pudl/package_data/meta/datapkg/datapackage.json:5153]: it's the attention line so like who a letter should be addressed to in the body....not sure why it's separate from `contact_firstname`, but it is.

[the remark this answers is not in the document]
Comment (aesharpe) [reply src/pudl/package_data/meta/datapkg/datapackage.json:5153]: I'll change it to `utility_attention_line` to it's more clear

**Pull request merged**